[ARM32/RyuJIT] Lowering: split struct argument less than 16 bytes

hseok-oh opened the pull request:
Enable passing split struct (less than 16 bytes) 
- morphing phase
- Lowering phase

Block at LSRA phase
- add NYI_ARM in TreeNodeInfoInitPutArgSplit()

Print split struct

related issue: #10722

hseok-oh commented:
cc/ @dotnet/arm32-contrib 

Example code
```
using System;
using System.Runtime.CompilerServices;

public struct IntStruct
{
    public int num1;
    public int num2;
    public int num3;
    public int num4;
}

public class BringUpTest
{
    const int Pass = 100;
    const int Fail = -1;

    [MethodImplAttribute(MethodImplOptions.NoInlining)]
    public static int StructAdd(int a, int b, IntStruct rp, int c, int d)
    {   
        return a + b + rp.num1 + rp.num2 + rp.num3 + rp.num4 + c + d;
    }

    public static int Main()
    {   
        IntStruct a = new IntStruct();

        a.num1 = 11;
        a.num2 = 12;
        a.num3 = 13;
        a.num4 = 14;

        int y = StructAdd(1, 2, a, 3, 4);
        if (y == 60) return Pass;
        else return Fail;
    }
}
```

Generated code: call `StructAdd`
```
IN0001: 00000A      movs    r2, 11
IN0002: 00000C      movs    r3, 12
IN0003: 00000E      movs    r0, 13
IN0004: 000010      movs    r1, 14
IN0005: 000012      str     r0, [sp]
IN0006: 000014      str     r1, [sp+0x04]
IN0007: 000016      movs    r0, 3
IN0008: 000018      str     r0, [sp+0x08]   
IN0009: 00001A      movs    r0, 4
IN000a: 00001C      str     r0, [sp+0x0c]   
IN000b: 00001E      movs    r0, 1
IN000c: 000020      movs    r1, 2
IN000d: 000022      movw    lr, 0xa930
IN000e: 000026      movt    lr, 0x7501
IN000f: 00002A      ldr     lr, [lr]
IN0010: 00002E      blx     lr          // BringUpTest:StructAdd(int,int,struct,int,int):int
```

BruceForstall commented:
cc @dotnet/jit-contrib

hqueue commented on src/jit/morph.cpp:
```diff
@@ -4760,6 +4754,24 @@ GenTreePtr Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPtr f
+            NYI_ARM("Struct split between integer registers and stack");
+        }
+    }
+    else if (fgEntryPtr->numSlot > 4)
```
CI failures are due to this compile-time error. maybe `numSlots` ?

hseok-oh replied:
You're right. Thanks.

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -788,6 +792,80 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
+                        NYI_ARM("lower: struct argument by fast tail call");
+                    }
+
+                    // Split FieldList: Two FieldLists for registers and stack
```
Wouldn't it make sense to leave the field list as-is, and add a flag to `GenTreeFieldList` to indicate whether it goes on the stack or in a register?

hseok-oh replied:
`GenTreePutArgSplit` has two operands: one is for register and other is for stack. `op1` references GenTreeFieldList passed by registers and each GenTreeFieldList reference `GenTreePutArgReg`.`op2` reference `GenTreePutArgStk` and operand of `GenTreePutArgStk` references `GenTreeFieldList` passed on stack. To avoid additional implementation for register allocation & code generation, I think split `GenTreeFieldList` is better.

CarolEidt commented on src/jit/lir.cpp:
```diff
@@ -1499,13 +1499,19 @@ bool LIR::Range::CheckLIR(Compiler* compiler, bool checkUnusedValues) const
+                assert(((node->OperGet() == GT_CALL) &&
+                        (def->OperIsStore() || (def->OperGet() == GT_PUTARG_STK) || (def->OperGet() == GT_NOP))) ||
+                       (node->OperGet() == GT_PUTARG_SPLIT && def->OperGet() == GT_PUTARG_STK));
+#else
```
It would seem to me that `GT_PUTARG_SPLIT` should always produce a value, since it defines at least one register.

hseok-oh replied:
This conditions for assertion needs because `node: GT_PUTARG_SPLIT` produce a value, but `def: GT_PUTARG_STK` does not, and `GenTreePutArgSplit` can have `GenTreePutArgStk` as child(operand).

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -788,6 +792,80 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
+
+            assert(fieldListPtr->IsFieldListHead());
+
+            for (unsigned ctr = 0; fieldListPtr != nullptr; fieldListPtr = fieldListPtr->Rest(), ctr++)
```
Please use a better name than `ctr` - in this case I actually think `fieldIndex` would be more meaningful.

hseok-oh replied:
I'll fix it. Thanks.

CarolEidt commented on src/jit/lsraarm.cpp:
```diff
@@ -717,6 +717,11 @@ void Lowering::TreeNodeInfoInit(GenTree* tree)
 
+        case GT_PUTARG_SPLIT:
+            info->srcCount = 0;
+            info->dstCount = 0;
```
How is the register modeled for this? Either this should produce multiple registers, that are then consumed by the call, OR the register fields should be contained, and consumed by the call. Either way, I think that this needs additional handling.

hseok-oh replied:
`GT_PUTARG_SPLIT` does not generate any code. So there's not need srcCount and dstCount. Registers are produced by `GT_PUTARG_REG`: the operand of `FIELD_LIST` and this `FIELD_LIST` is operand of `GT_PUTARG_SPLIT`.

BruceForstall commented:
In the "lowering call (before)" output, I see:
```
                                                  /--*  t77    void   arg2 out+00
                                                  +--*  t33    int    arg3 out+08
                                                  +--*  t34    int    arg4 out+0c
                                                  +--*  t30    int    arg0 in r0
                                                  +--*  t31    int    arg1 in r1
N022 ( 36, 30) [000035] --C-G--------       t35 = *  call      int    BringUpTest.StructAdd $146
```
This seems somewhat confusing, since it says `t77    void   arg2 out+00`. But t77 will be split between registers and stack. It's also confusing because t77 represents a 16-byte struct, but arg3 is only +8 bytes from arg2.

Can we improve the dump so it more clearly shows where the argument is going?

I presume this is determined by fgMorphArgs(), long before Lowering.

hqueue commented:
@dotnet-bot please test Tizen armel Cross Debug Build

hseok-oh commented:
@BruceForstall I change the log to print register and stack for split struct.
```
N017 (  1,  1) [000025] -------------       t25 =    const     int    2 $48
                                                  /--*  t67    void   arg2 r2,r3 out+00
                                                  +--*  t27    int    arg3 out+04
                                                  +--*  t28    int    arg4 out+08
                                                  +--*  t24    int    arg0 in r0
                                                  +--*  t25    int    arg1 in r1
N020 ( 34, 27) [000029] --C-G--------       t29 = *  call      int    BringUpTest.StructAdd $145
```
But this makes some formatting problem at other places, because the message for argument is too long.

```
fgMorphTree BB01, stmt 5 (after)
               [000039] --C-G+-------             *  jmpTrue   void  
               [000037] -----+-------             |  /--*  const     int    46
               [000038] N-C-G+-N-US--             \--*  !=        int
               [000029] --C-G+-------                \--*  call      int    BringUpTest.StructAdd
               [000071] -------------                   |     /--*  <fldList> void   int at offset 8
               [000070] -------------                   |     |  \--*  lclVar    int    V03 tmp2
               [000069] -------------                   |  /--*  <fldList> void   int at offset 4
               [000068] -------------                   |  |  \--*  lclVar    int    V02 tmp1
               [000067] ------------- arg2 r2,r3 out+00      +--*  <fldList> void   int at offset 0
               [000066] -------------                   |  \--*  lclVar    int    V01 tmp0
               [000027] -----+------- arg3 out+04       +--*  const     int    3
               [000028] -----+------- arg4 out+08       +--*  const     int    4
               [000024] -----+------- arg0 in r0        +--*  const     int    1
               [000025] -----+------- arg1 in r1        \--*  const     int    2
```

To avoid this, it needs to change default message length for this: from 25 to 31
https://github.com/dotnet/coreclr/blob/master/src/jit/gentree.cpp#L10024
 But I don't know it's ok to change this value.

hseok-oh commented:
@dotnet-bot test Tizen armel Cross Debug Build please

BruceForstall commented:
> To avoid this, it needs to change default message length for this: from 25 to 31
https://github.com/dotnet/coreclr/blob/master/src/jit/gentree.cpp#L10024
But I don't know it's ok to change this value.

I'm ok with you changing this value, but do it only for `defined(_TARGET_ARM_) && !defined(LEGACY_BACKEND)`

BruceForstall commented:
@hseok-oh What is the relationship between this PR and #12050? Is this one obsolete, subsumed (replaced) by #12050? If so, can it be closed?

hseok-oh commented:
@BruceForstall @CarolEidt I rebased and updated. PTAL

Enable passing split struct (less than 16 bytes)
- morphing phase
- Lowering phase

Block at LSRA phase
- add NYI_ARM in `TreeNodeInfoInitPutArgSplit()`

Print split struct

hseok-oh commented:
Morphing
```
fgMorphTree BB01, stmt 6 (before)
               [000045] --C-G--------             *  JTRUE     void
               [000043] -------------             |  /--*  CNS_INT   int    60
               [000044] --C-G--------             \--*  EQ        int
               [000035] --C-G--------                \--*  CALL      int    BringUpTest.StructAdd
               [000030] ------------- arg0              +--*  CNS_INT   int    1
               [000031] ------------- arg1              +--*  CNS_INT   int    2
               [000039] ------------- arg2              +--*  OBJ(16)   struct
               [000038] L------------                   |  \--*  ADDR      byref
               [000032] -------------                   |     \--*  LCL_VAR   struct(P) V00 loc0
                                                        |     \--*    int    V00.num1 (offs=0x00) -> V01 tmp0
                                                        |     \--*    int    V00.num2 (offs=0x04) -> V02 tmp1
                                                        |     \--*    int    V00.num3 (offs=0x08) -> V03 tmp2
                                                        |     \--*    int    V00.num4 (offs=0x0c) -> V04 tmp3
               [000033] ------------- arg3              +--*  CNS_INT   int    3
               [000034] ------------- arg4              \--*  CNS_INT   int    4
```
```
fgMorphTree BB01, stmt 6 (after)
               [000045] --C-G+-------             *  JTRUE     void  
               [000043] -----+-------             |  /--*  CNS_INT   int    60
               [000044] J-C-G+-N--S--             \--*  EQ        int
               [000035] --C-G+-------                \--*  CALL      int    BringUpTest.StructAdd
               [000033] -----+------- arg3 out+08       +--*  CNS_INT   int    3
               [000034] -----+------- arg4 out+0c       +--*  CNS_INT   int    4
               [000085] -------------                   |        /--*  FIELD_LIST void   int at offset 12
               [000084] -------------                   |        |  \--*  LCL_VAR   int    V04 tmp3
               [000083] -------------                   |     /--*  FIELD_LIST void   int at offset 8
               [000082] -------------                   |     |  \--*  LCL_VAR   int    V03 tmp2
               [000081] -------------                   |  /--*  FIELD_LIST void   int at offset 4
               [000080] -------------                   |  |  \--*  LCL_VAR   int    V02 tmp1
               [000079] ------------- arg2 r2,r3 out+00      +--*  FIELD_LIST void   int at offset 0
               [000078] -------------                   |  \--*  LCL_VAR   int    V01 tmp0
               [000030] -----+------- arg0 in r0        +--*  CNS_INT   int    1
               [000031] -----+------- arg1 in r1        \--*  CNS_INT   int    2
```

Lowering
```
lowering call (before):
N004 (  1,  1) [000033] -------------       t33 =    CNS_INT   int    3 $46
N005 (  1,  1) [000034] -------------       t34 =    CNS_INT   int    4 $47
N011 (  1,  1) [000078] -------------       t78 =    CNS_INT   int    11 $42
N012 (  1,  1) [000080] -------------       t80 =    CNS_INT   int    12 $43
N013 (  1,  1) [000082] -------------       t82 =    CNS_INT   int    13 $44
N014 (  1,  1) [000084] -------------       t84 =    CNS_INT   int    14 $45
                                                  /--*  t78    int
                                                  +--*  t80    int
                                                  +--*  t82    int
                                                  +--*  t84    int
N018 (  7, 11) [000079] -------------       t79 = *  FIELD_LIST void   int at offset 0 $1c8
N019 (  1,  1) [000030] -------------       t30 =    CNS_INT   int    1 $48
N020 (  1,  1) [000031] -------------       t31 =    CNS_INT   int    2 $49
                                                  /--*  t33    int    arg3 out+08
                                                  +--*  t34    int    arg4 out+0c
                                                  +--*  t79    void   arg2 r2,r3 out+00
                                                  +--*  t30    int    arg0 in r0
                                                  +--*  t31    int    arg1 in r1
N024 ( 33, 30) [000035] --C-G--------       t35 = *  CALL      int    BringUpTest.StructAdd $146
```
```
lowering call (after):
N004 (  1,  1) [000033] -------------       t33 =    CNS_INT   int    3 $46
                                                  /--*  t33    int
               [000086] -------------             *  PUTARG_STK [+0x08] int
N005 (  1,  1) [000034] -------------       t34 =    CNS_INT   int    4 $47
                                                  /--*  t34    int
               [000087] -------------             *  PUTARG_STK [+0x0c] int
N011 (  1,  1) [000078] -------------       t78 =    CNS_INT   int    11 $42
N012 (  1,  1) [000080] -------------       t80 =    CNS_INT   int    12 $43
N013 (  1,  1) [000082] -------------       t82 =    CNS_INT   int    13 $44
N014 (  1,  1) [000084] -------------       t84 =    CNS_INT   int    14 $45
                                                  /--*  t78    int
                                                  +--*  t80    int
                                                  +--*  t82    int
                                                  +--*  t84    int
N018 (  7, 11) [000079] -------------       t79 = *  FIELD_LIST void   int at offset 0 $1c8
                                                  /--*  t79    void
               [000088] -------------       t88 = *  PUTARG_SPLIT struct
N019 (  1,  1) [000030] -------------       t30 =    CNS_INT   int    1 $48
                                                  /--*  t30    int
               [000089] -------------       t89 = *  PUTARG_REG int
N020 (  1,  1) [000031] -------------       t31 =    CNS_INT   int    2 $49
                                                  /--*  t31    int
               [000090] -------------       t90 = *  PUTARG_REG int
N001 (  3,  7) [000091] -------------       t91 =    CNS_INT(h) int    0x750494C8 ftn REG NA
                                                  /--*  t91    int
N002 (  6,  9) [000092] -------------       t92 = *  IND       int
                                                  /--*  t88    struct arg2 r2,r3 out+00
                                                  +--*  t89    int    arg0 in r0
                                                  +--*  t90    int    arg1 in r1
                                                  +--*  t92    int    control expr
N024 ( 33, 30) [000035] --C-G--------       t35 = *  CALL      int    BringUpTest.StructAdd $146

```

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -831,26 +831,44 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
     if (info->isSplit)
     {
-        if (arg->OperGet() != GT_OBJ)
+        assert(arg->OperGet() != GT_OBJ || arg->OperGet() != GT_FIELD_LIST);
```
This assert doesn't make sense. It will always be true. I think you meant to use `==` in these.

hseok-oh replied:
I'll fix it.

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -831,26 +831,44 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
-        BYTE*               gcLayout = new (comp, CMK_Codegen) BYTE[info->numSlots + info->numRegs];
-        unsigned            numRefs  = comp->info.compCompHnd->getClassGClayout(arg->gtObj.gtClass, gcLayout);
-        argSplit->setGcPointers(numRefs, gcLayout);
+        if (arg->OperGet() != GT_FIELD_LIST)
```
I'm curious why the gc info hasn't been initialized on the `GT_OBJ`, but at least you could check `arg->AsObj->IsGCInfoInitialized()` first.

hseok-oh replied:
I'll update it to use assigned gc info on the GT_OBJ if it is initialized.

CarolEidt commented on src/jit/lsraarmarch.cpp:
```diff
@@ -771,6 +771,11 @@ void Lowering::TreeNodeInfoInitPutArgSplit(GenTreePutArgSplit* argNode, TreeNode
 
+    if (putArgChild->OperGet() == GT_FIELD_LIST)
+    {
+        NYI_ARM("LSRA: Oper for split struct argument is GT_FIELD_LIST");
```
If I understand correctly, this is just because you haven't gotten to this step yet?

hseok-oh replied:
Exactly. I'll have a plan to make new PRs for lsra and codegen step.

hseok-oh commented:
@dotnet-bot test Ubuntu arm Cross Release Build please

hseok-oh commented:
@CarolEidt I've applied your feedback. PTAL.
- Fix assert bug in Lowering::NewPutArg
- Use assigned gc info on the GT_OBJ if it is initialized

CarolEidt commented:
@BruceForstall - you added the "No Merge" label some time ago - do you still feel that it should not be merged?
